napari_cellseg3d/code_models/models/wnet/model.py

```diff
@@ -0,0 +1,234 @@
+        super(UNet, self).__init__()
+        # self.device = device
+        self.channels = channels
+        self.max_pool = nn.MaxPool3d(2)
+        self.in_b = InBlock(in_channels, self.channels[0], dropout=dropout)
+        self.conv1 = Block(channels[0], self.channels[1], dropout=dropout)
+        self.conv2 = Block(channels[1], self.channels[2], dropout=dropout)
+        self.conv3 = Block(channels[2], self.channels[3], dropout=dropout)
+        self.bot = Block(channels[3], self.channels[4], dropout=dropout)
+        self.deconv1 = Block(channels[4], self.channels[3], dropout=dropout)
+        self.conv_trans1 = nn.ConvTranspose3d(
+            self.channels[4], self.channels[3], 2, stride=2
+        )
+        self.deconv2 = Block(channels[3], self.channels[2], dropout=dropout)
+        self.conv_trans2 = nn.ConvTranspose3d(
+            self.channels[3], self.channels[2], 2, stride=2
+        )
+        self.deconv3 = Block(channels[2], self.channels[1], dropout=dropout)
+        self.conv_trans3 = nn.ConvTranspose3d(
+            self.channels[2], self.channels[1], 2, stride=2
+        )
+        self.out_b = OutBlock(channels[1], out_channels, dropout=dropout)
+        self.conv_trans_out = nn.ConvTranspose3d(
+            self.channels[1], self.channels[0], 2, stride=2
+        )
+
+        self.sm = nn.Softmax(dim=1)
+        self.encoder = encoder
+
+    def forward(self, x):
+        """Forward pass of the U-Net model."""
+        in_b = self.in_b(x)
+        c1 = self.conv1(self.max_pool(in_b))
+        c2 = self.conv2(self.max_pool(c1))
+        c3 = self.conv3(self.max_pool(c2))
+        x = self.bot(self.max_pool(c3))
+        x = self.deconv1(
+            torch.cat(
+                [
+                    c3,
+                    self.conv_trans1(x),
+                ],
+                dim=1,
+            )
+        )
+        x = self.deconv2(
+            torch.cat(
+                [
+                    c2,
+                    self.conv_trans2(x),
+                ],
+                dim=1,
+            )
+        )
+        x = self.deconv3(
+            torch.cat(
+                [
+                    c1,
+                    self.conv_trans3(x),
+                ],
+                dim=1,
+            )
+        )
+        x = self.out_b(
+            torch.cat(
+                [
+                    in_b,
+                    self.conv_trans_out(x),
+                ],
+                dim=1,
+            )
+        )
+        if self.encoder:
+            x = self.sm(x)
+        return x
+
+
+class InBlock(nn.Module):
+    """Input block of the U-Net architecture."""
+
+    def __init__(self, in_channels, out_channels, dropout=0.65):
+        super(InBlock, self).__init__()
+        # self.device = device
+        self.module = nn.Sequential(
+            nn.Conv3d(in_channels, out_channels, 3, padding=1),
+            nn.ReLU(),
+            nn.Dropout(p=dropout),
+            nn.BatchNorm3d(out_channels),
```

Comment:
I wonder why they use BatchNorm instead of Layer or GroupNorm. Do you know if they made any ablation studies on that?